utils/extraction.py

```diff
@@ -43,23 +78,105 @@ def extract_preview_data(
+        row: Dict[str, Any] = {"key": str(i)}
         for j, col in enumerate(df.columns):
-            row[str(j)] = parse(df.loc[i, col], coltypes[j])
+            parsed = parse(df.loc[i, col], coltypes[j])
+            should_strip = coltypes[j] == ColumnTypes.STRING and trimWhitespaces
+            row[str(j)] = str(parsed).strip() if should_strip else parsed
         return row
 
     rows = [get_ith_row_from_df(i) for i in range(len(df))]
 
-    return {
+    extracted: ExtractedData = {
         "rows": rows,
         "columns": columns,
-    }, None
+        "extra_headers": extra_headers,
+        "column_stats": calculate_column_stats(df_dict, coltypes)
+        if calculate_stats
+        else None,
+    }
+    return extracted, None
+
+
+def extract_extra_headers(
+    xl: pd.ExcelFile, sheetname: str, header_level: int
+) -> List[List[str]]:
+    if header_level < 1:
+        # There's nothing to read before header 1
+        return []
+    df: pd.DataFrame = cast(
+        pd.DataFrame,
+        xl.parse(
+            sheetname,
+            nrows=header_level,
+            header=None,
+        ),
+    )
+    return [[str(e) for e in row] for row in df.itertuples(index=False)]
+
+
+def calculate_column_stats(df_dict: dict, coltypes: Dict[int, ColumnTypes]):
+    """
+    Calculate column wise stats.
+    df_dict is a dict of the form:
+        {
+            col1: [item1, item2, ...],
+            col2: [item1, item2, ...],
+        }
+    which is the result of dataframe.to_dict()
+    """
+    column_stats = []
+    for colname, coltype in zip(df_dict.keys(), coltypes.values()):
+        if coltype == ColumnTypes.INTEGER:
+            column_stats.append(calculate_stats_for_numeric_col(df_dict[colname]))
+        elif coltype == ColumnTypes.FLOATING:
+            column_stats.append(calculate_stats_for_numeric_col(df_dict[colname]))
+        else:
+            column_stats.append(calculate_stats_for_string_col(df_dict[colname]))
+    return column_stats
+
+
+def calculate_stats_for_numeric_col(items: list):
+    # TODO: optimize the list(use np/pd). But this should happen from the extraction phase itself
+    return {
+        "min": float(np.min(items)),
+        "max": float(np.max(items)),
+        "mean": float(np.mean(items)),
+        "median": float(np.median(items)),
+        "std_deviation": float(np.std(items)),
+        "total_count": len(items),
+        "na_count": len([x for x in items if x is None]),
+    }
+
+
+def calculate_stats_for_string_col(items: list):
+    max_len = 0
+    min_len = INFINITY
+    for x in items:
+        if x is None:
+            continue
+        length = len(x)
+        if length > max_len:
+            max_len = length
+        if length < min_len:
+            min_len = length
+
+    # TODO: optimize the following calculations
+    return {
+        "total_count": len(items),
+        "na_count": len([x for x in items if x is None]),
+        "unique_count": len(set(items)),
```

Comment:
@bewakes how about using numpy for all starts calculations?

Reply:
That's the ultimate plan. The calculations themselves need to be optimized and are naively done at the moment. We'll work on these after the first MVP.